packages/react-i18n/src/utilities/translate.tsx

```diff
@@ -193,104 +202,106 @@ function translateWithDictionary(
-  const allReplacementsArePrimitives = Object.keys(replacements).every(
-    (key) => typeof replacements[key] !== 'object',
-  );
-
-  if (allReplacementsArePrimitives) {
-    return str.replace(REPLACE_REGEX, (match) => {
-      const replacement = match.substring(1, match.length - 1);
-
-      if (!Object.prototype.hasOwnProperty.call(replacements, replacement)) {
-        throw new MissingReplacementError(replacement, replacements);
+  const pieces: (string | React.ReactElement<any>)[] = [];
+
+  const replaceFinder = new RegExp(interpolate || DEFAULT_INTERPOLATION, 'g');
+
+  let matchIndex = 0;
+  let lastOffset = 0;
+
+  // Uses replace callback, but not its return value
+  str.replace(
+    replaceFinder,
+    (
+      match,
+      _fullPlaceholder: string,
+      replacementKey: string,
+      offset: number,
+    ) => {
+      if (!replacementKey) return '';
+
+      if (!Object.prototype.hasOwnProperty.call(replacements, replacementKey)) {
+        throw new MissingReplacementError(replacementKey, replacements);
       }
 
-      return replacements[replacement] as string;
-    });
-  } else {
-    const pieces: (string | React.ReactElement<any>)[] = [];
-
-    let match = replaceFinder.exec(str);
-    let matchIndex = 0;
+      matchIndex += 1;
 
-    while (match) {
-      const regularText = match[1];
-      const replacement = match[3];
+      const replacementValue = replacements[replacementKey];
+      const finalReplacement =
+        replacementValue && React.isValidElement(replacementValue)
+          ? React.cloneElement(replacementValue, {key: matchIndex})
+          : (replacementValue as string);
 
-      if (match.index >= str.length) {
-        break;
-      }
-
-      if (regularText) {
-        pieces.push(regularText);
-      }
+      // Push the previous part if it exists
+      const previousString = str.substring(lastOffset, offset);
+      if (previousString) pieces.push(previousString);
 
-      if (replacement) {
-        if (!Object.prototype.hasOwnProperty.call(replacements, replacement)) {
-          throw new MissingReplacementError(replacement, replacements);
-        }
+      // Push the new part with the replacement
+      pieces.push(finalReplacement);
 
-        matchIndex += 1;
-        const finalReplacement = React.isValidElement(replacements[replacement])
-          ? React.cloneElement(
-              replacements[replacement] as React.ReactElement<any>,
-              {key: matchIndex},
-            )
-          : (replacements[replacement] as string);
+      lastOffset = offset + match.length;
 
-        pieces.push(finalReplacement);
-      }
-
-      match = replaceFinder.exec(str);
-    }
+      // to satisfy the typechecker
+      return '';
+    },
+  );
 
-    replaceFinder.lastIndex = 0;
+  // Push the last part of the source string
+  pieces.push(str.substr(lastOffset));
 
-    return pieces;
-  }
+  return pieces.every(isString) ? pieces.join('') : pieces;
```

Comment:
This line replaces the `allReplacementsArePrimitives` value and its related logic.